kroxylicious/src/main/java/io/kroxylicious/proxy/codec/KafkaRequestEncoder.java

```diff
@@ -16,15 +16,120 @@
  */
 package io.kroxylicious.proxy.codec;
 
+import org.apache.kafka.common.message.ProduceRequestData;
+import org.apache.kafka.common.protocol.ApiKeys;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelHandlerContext;
+
 public class KafkaRequestEncoder extends KafkaMessageEncoder<RequestFrame> {
     private static final Logger LOGGER = LogManager.getLogger(KafkaRequestEncoder.class);
+    public static final int LENGTH = 4;
+    public static final int API_KEY = 2;
+    public static final int API_VERSION = 2;
+    private final CorrelationManager correlationManager;
+
+    public KafkaRequestEncoder(CorrelationManager correlationManager) {
+        this.correlationManager = correlationManager;
+    }
 
     @Override
     protected Logger log() {
         return LOGGER;
     }
 
+    @Override
+    protected void encode(ChannelHandlerContext ctx, RequestFrame frame, ByteBuf out) throws Exception {
+        super.encode(ctx, frame, out);
+        // TODO re-reading from the encoded buffer like this is ugly
+        // probably better to just include apiKey and apiVersion in the frame
+        var ri = out.readerIndex();
```

Comment:
You can just read/write what you need by specifying absolute offsets

Reply:
True, I hadn't noticed the `get*` and `set*` methods. But it doesn't work so well for varint encodings which need to read a variable number of bytes (see the `while` loop in `ByteBufAccessor#readUnsignedVarint`, for example). TBH I don't see how to encapsulate reading a varint using absolute offsets because there is no easy way to increment the index in a caller or `readUnsignedVarint` while also returning the `int` that was read.